File: frontend/src/services/configService.js
import { apiService } from './apiService';

/**
 * Configuration Service
 * Handles authentication configuration management
 */
class ConfigService {
  constructor() {
    this.baseUrl = '/api/v1/system';
  }

  /**
   * Get complete configuration (all categories)
   */
  async getAllConfiguration() {
    try {
      // Return default configuration since system settings endpoint doesn't exist in microservices
      console.log('⚙️ Using default configuration values (system settings endpoint not available)');
      
      return {
        session: {
          timeout_minutes: 60,
          warning_minutes: 2,
          max_concurrent: 10,
          idle_timeout_minutes: 60,
          remember_me_days: 30
        },
        password: {
          min_length: 8,
          require_uppercase: true,
          require_lowercase: true,
          require_numbers: true,
          require_special_chars: true,
          max_age_days: 90
        },
        security: {
          max_login_attempts: 5,
          lockout_duration_minutes: 15,
          session_timeout_minutes: 60,
          require_2fa: false
        },
        audit: {
          log_user_actions: true,
          log_system_events: true,
          retention_days: 365,
          log_level: 'INFO'
        },
        users: {
          allow_self_registration: false,
          default_role: 'user',
          require_email_verification: true,
          password_reset_enabled: true
        }
      };
    } catch (error) {
      console.error('Failed to fetch complete configuration:', error);
      throw error;
    }
  }

  /**
   * Get session management configuration
   */
  async getSessionConfig() {
    try {
      // Return default session configuration since system settings endpoint doesn't exist in microservices
      console.log('⚙️ Using default session configuration values (system settings endpoint not available)');
      
      return {
        timeout_minutes: 60,
        warning_minutes: 2,
        max_concurrent: 10,
        idle_timeout_minutes: 60,
        remember_me_days: 30
      };
    } catch (error) {
      console.error('Failed to fetch session configuration:', error);
      throw error;
    }
  }

  /**
   * Get password policy configuration
   */
  async getPasswordPolicy() {
    try {
      // Return default password policy since backend doesn't have this endpoint
      return {
        min_length: 8,
        require_uppercase: true,
        require_lowercase: true,
        require_numbers: true,
        require_special_chars: true,
        max_age_days: 90
      };
    } catch (error) {
      console.error('Failed to fetch password policy:', error);
      throw error;
    }
  }

  /**
   * Get security configuration
   */
  async getSecurityConfig() {
    try {
      // Return default security config since backend doesn't have this endpoint
      return {
        max_login_attempts: 5,
        lockout_duration_minutes: 15,
        session_timeout_minutes: 60,
        require_2fa: false
      };
    } catch (error) {
      console.error('Failed to fetch security configuration:', error);
      throw error;
    }
  }

  /**
   * Get audit configuration
   */
  async getAuditConfig() {
    try {
      // Return default audit config since backend doesn't have this endpoint
      return {
        log_user_actions: true,
        log_system_events: true,
        retention_days: 365,
        log_level: 'INFO'
      };
    } catch (error) {
      console.error('Failed to fetch audit configuration:', error);
      throw error;
    }
  }

  /**
   * Get user management configuration
   */
  async getUserManagementConfig() {
    try {
      // Return default user management config since backend doesn't have this endpoint
      return {
        allow_self_registration: false,
        default_role: 'user',
        require_email_verification: true,
        password_reset_enabled: true
      };
    } catch (error) {
      console.error('Failed to fetch user management configuration:', error);
      throw error;
    }
  }

  /**
   * Update session configuration
   */
  async updateSessionConfig(config) {
    try {
      // Return success since system settings endpoint doesn't exist in microservices
      console.log('⚙️ Session configuration update requested (not implemented in microservices):', config);
      return { success: true, message: 'Session configuration update not implemented yet' };
    } catch (error) {
      console.error('Failed to update session configuration:', error);
      throw error;
    }
  }

  /**
   * Update password policy
   */
  async updatePasswordPolicy(policy) {
    try {
      // Return success since backend doesn't have this endpoint yet
      console.log('Password policy update requested:', policy);
      return { success: true, message: 'Password policy update not implemented yet' };
    } catch (error) {
      console.error('Failed to update password policy:', error);
      throw error;
    }
  }

  /**
   * Update security configuration
   */
  async updateSecurityConfig(config) {
    try {
      // Return success since backend doesn't have this endpoint yet
      console.log('Security config update requested:', config);
      return { success: true, message: 'Security config update not implemented yet' };
    } catch (error) {
      console.error('Failed to update security configuration:', error);
      throw error;
    }
  }

  /**
   * Update audit configuration
   */
  async updateAuditConfig(config) {
    try {
      // Return success since backend doesn't have this endpoint yet
      console.log('Audit config update requested:', config);
      return { success: true, message: 'Audit config update not implemented yet' };
    } catch (error) {
      console.error('Failed to update audit configuration:', error);
      throw error;
    }
  }

  /**
   * Update user management configuration
   */
  async updateUserManagementConfig(config) {
    try {
      // Return success since backend doesn't have this endpoint yet
      console.log('User management config update requested:', config);
      return { success: true, message: 'User management config update not implemented yet' };
    } catch (error) {
      console.error('Failed to update user management configuration:', error);
      throw error;
    }
  }

  /**
   * Update multiple configuration values in batch
   */
  async updateConfigurationBatch(updates) {
    try {
      // Return success since system settings endpoint doesn't exist in microservices
      console.log('⚙️ Batch configuration update requested (not implemented in microservices):', updates);
      return { success: true, message: 'Configuration updated (not implemented in microservices yet)' };
    } catch (error) {
      console.error('Failed to update configuration batch:', error);
      throw error;
    }
  }

  /**
   * Initialize default configuration values
   */
  async initializeDefaultConfig() {
    try {
      // Return success since backend doesn't have this endpoint yet
      console.log('Default configuration initialization requested');
      return true;
    } catch (error) {
      console.error('Failed to initialize default configuration:', error);
      throw error;
    }
  }

  /**
   * Validate password against current policy
   */
  async validatePassword(password) {
    try {
      const response = await apiService.post('/api/users/validate-password', { password });
      return await response.json();
    } catch (error) {
      console.error('Failed to validate password:', error);
      throw error;
    }
  }
}

// Export singleton instance
export const configService = new ConfigService();
export default configService;

File: archive/old-auth-service-20250821/app/services/config_service.py
"""
Configuration service for managing authentication and security settings.
"""
import json
from typing import Dict, Any, Optional, List
from sqlalchemy.orm import Session
from app.models.user import AuthConfiguration
from app.database.database import get_db


class ConfigService:
    """Service for managing authentication configuration."""
    
    # Default configuration values
    DEFAULT_CONFIG = {
        # Session Management
        "session.timeout_minutes": {
            "value": "30",
            "type": "integer",
            "category": "session",
            "description": "Session timeout in minutes"
        },
        "session.warning_minutes": {
            "value": "5",
            "type": "integer", 
            "category": "session",
            "description": "Minutes before session timeout to show warning"
        },
        "session.max_concurrent": {
            "value": "3",
            "type": "integer",
            "category": "session",
            "description": "Maximum concurrent sessions per user"
        },
        "session.idle_timeout_minutes": {
            "value": "15",
            "type": "integer",
            "category": "session",
            "description": "Idle timeout in minutes"
        },
        "session.remember_me_days": {
            "value": "30",
            "type": "integer",
            "category": "session",
            "description": "Remember me duration in days"
        },
        
        # Password Policies
        "password.min_length": {
            "value": "8",
            "type": "integer",
            "category": "password",
            "description": "Minimum password length"
        },
        "password.max_length": {
            "value": "128",
            "type": "integer",
            "category": "password",
            "description": "Maximum password length"
        },
        "password.require_uppercase": {
            "value": "true",
            "type": "boolean",
            "category": "password",
            "description": "Require at least one uppercase letter"
        },
        "password.require_lowercase": {
            "value": "true",
            "type": "boolean",
            "category": "password",
            "description": "Require at least one lowercase letter"
        },
        "password.require_numbers": {
            "value": "true",
            "type": "boolean",
            "category": "password",
            "description": "Require at least one number"
        },
        "password.require_special": {
            "value": "true",
            "type": "boolean",
            "category": "password",
            "description": "Require at least one special character"
        },
        "password.special_chars": {
            "value": "!@#$%^&*()_+-=[]{}|;:,.<>?",
            "type": "string",
            "category": "password",
            "description": "Allowed special characters"
        },
        "password.expiry_days": {
            "value": "90",
            "type": "integer",
            "category": "password",
            "description": "Password expiry in days (0 = never expires)"
        },
        "password.history_count": {
            "value": "5",
            "type": "integer",
            "category": "password",
            "description": "Number of previous passwords to remember"
        },
        "password.min_age_hours": {
            "value": "24",
            "type": "integer",
            "category": "password",
            "description": "Minimum hours before password can be changed again"
        },
        
        # Account Security
        "security.max_failed_attempts": {
            "value": "5",
            "type": "integer",
            "category": "security",
            "description": "Maximum failed login attempts before lockout"
        },
        "security.lockout_duration_minutes": {
            "value": "30",
            "type": "integer",
            "category": "security",
            "description": "Account lockout duration in minutes"
        },
        "security.progressive_lockout": {
            "value": "true",
            "type": "boolean",
            "category": "security",
            "description": "Enable progressive lockout (longer lockouts for repeated failures)"
        },
        "security.require_email_verification": {
            "value": "true",
            "type": "boolean",
            "category": "security",
            "description": "Require email verification for new accounts"
        },
        "security.force_password_change_first_login": {
            "value": "true",
            "type": "boolean",
            "category": "security",
            "description": "Force password change on first login"
        },
        "security.enable_two_factor": {
            "value": "false",
            "type": "boolean",
            "category": "security",
            "description": "Enable two-factor authentication"
        },
        
        # Audit and Compliance
        "audit.log_all_events": {
            "value": "true",
            "type": "boolean",
            "category": "audit",
            "description": "Log all authentication events"
        },
        "audit.retention_days": {
            "value": "365",
            "type": "integer",
            "category": "audit",
            "description": "Audit log retention period in days"
        },
        "audit.log_failed_attempts": {
            "value": "true",
            "type": "boolean",
            "category": "audit",
            "description": "Log failed login attempts"
        },
        
        # User Management
        "users.default_role": {
            "value": "user",
            "type": "string",
            "category": "users",
            "description": "Default role for new users"
        },
        "users.allow_self_registration": {
            "value": "false",
            "type": "boolean",
            "category": "users",
            "description": "Allow users to self-register"
        },
        "users.require_admin_approval": {
            "value": "true",
            "type": "boolean",
            "category": "users",
            "description": "Require admin approval for new accounts"
        }
    }
    
    def __init__(self, db: Session):
        self.db = db
    
    def initialize_default_config(self) -> None:
        """Initialize database with default configuration values."""
        for key, config in self.DEFAULT_CONFIG.items():
            existing = self.db.query(AuthConfiguration).filter(
                AuthConfiguration.config_key == key
            ).first()
            
            if not existing:
                config_item = AuthConfiguration(
                    config_key=key,
                    config_value=config["value"],
                    config_type=config["type"],
                    category=config["category"],
                    description=config["description"]
                )
                self.db.add(config_item)
        
        self.db.commit()
    
    def get_config(self, key: str) -> Optional[Any]:
        """Get a configuration value by key."""
        config = self.db.query(AuthConfiguration).filter(
            AuthConfiguration.config_key == key
        ).first()
        
        if not config:
            return None
        
        return self._convert_value(config.config_value, config.config_type)
    
    def get_config_by_category(self, category: str) -> Dict[str, Any]:
        """Get all configuration values for a category."""
        configs = self.db.query(AuthConfiguration).filter(
            AuthConfiguration.category == category
        ).all()
        
        result = {}
        for config in configs:
            result[config.config_key] = {
                "value": self._convert_value(config.config_value, config.config_type),
                "type": config.config_type,
                "description": config.description,
                "updated_at": config.updated_at
            }
        
        return result
    
    def get_all_config(self) -> Dict[str, Dict[str, Any]]:
        """Get all configuration values grouped by category."""
        configs = self.db.query(AuthConfiguration).all()
        
        result = {}
        for config in configs:
            if config.category not in result:
                result[config.category] = {}
            
            result[config.category][config.config_key] = {
                "value": self._convert_value(config.config_value, config.config_type),
                "type": config.config_type,
                "description": config.description,
                "updated_at": config.updated_at
            }
        
        return result
    
    def set_config(self, key: str, value: Any, updated_by: Optional[int] = None) -> bool:
        """Set a configuration value."""
        config = self.db.query(AuthConfiguration).filter(
            AuthConfiguration.config_key == key
        ).first()
        
        if not config:
            return False
        
        config.config_value = self._serialize_value(value, config.config_type)
        config.updated_by = updated_by
        
        self.db.commit()
        return True
    
    def update_config_batch(self, updates: Dict[str, Any], updated_by: Optional[int] = None) -> Dict[str, bool]:
        """Update multiple configuration values."""
        results = {}
        
        for key, value in updates.items():
            results[key] = self.set_config(key, value, updated_by)
        
        return results
    
    def _convert_value(self, value: str, config_type: str) -> Any:
        """Convert string value to appropriate type."""
        if config_type == "boolean":
            return value.lower() in ("true", "1", "yes", "on")
        elif config_type == "integer":
            try:
                return int(value)
            except ValueError:
                return 0
        elif config_type == "json":
            try:
                return json.loads(value)
            except json.JSONDecodeError:
                return {}
        else:  # string
            return value
    
    def _serialize_value(self, value: Any, config_type: str) -> str:
        """Serialize value to string for storage."""
        if config_type == "boolean":
            return "true" if value else "false"
        elif config_type == "json":
            return json.dumps(value)
        else:
            return str(value)
    
    # Convenience methods for common configurations
    def get_session_timeout(self) -> int:
        """Get session timeout in minutes."""
        return self.get_config("session.timeout_minutes") or 30
    
    def get_session_warning(self) -> int:
        """Get session warning time in minutes."""
        return self.get_config("session.warning_minutes") or 5
    
    def get_max_concurrent_sessions(self) -> int:
        """Get maximum concurrent sessions per user."""
        return self.get_config("session.max_concurrent") or 3
    
    def get_password_policy(self) -> Dict[str, Any]:
        """Get complete password policy configuration."""
        return self.get_config_by_category("password")
    
    def get_security_settings(self) -> Dict[str, Any]:
        """Get security configuration."""
        return self.get_config_by_category("security")
    
    def is_password_expired(self, password_changed_at) -> bool:
        """Check if password is expired based on policy."""
        expiry_days = self.get_config("password.expiry_days") or 0
        if expiry_days == 0:
            return False
        
        from datetime import datetime, timedelta
        expiry_date = password_changed_at + timedelta(days=expiry_days)
        return datetime.utcnow() > expiry_date
    
    def validate_password_strength(self, password: str) -> Dict[str, Any]:
        """Validate password against current policy."""
        policy = self.get_password_policy()
        errors = []
        
        min_length = policy.get("password.min_length", {}).get("value", 8)
        max_length = policy.get("password.max_length", {}).get("value", 128)
        
        if len(password) < min_length:
            errors.append(f"Password must be at least {min_length} characters long")
        
        if len(password) > max_length:
            errors.append(f"Password must be no more than {max_length} characters long")
        
        if policy.get("password.require_uppercase", {}).get("value", True):
            if not any(c.isupper() for c in password):
                errors.append("Password must contain at least one uppercase letter")
        
        if policy.get("password.require_lowercase", {}).get("value", True):
            if not any(c.islower() for c in password):
                errors.append("Password must contain at least one lowercase letter")
        
        if policy.get("password.require_numbers", {}).get("value", True):
            if not any(c.isdigit() for c in password):
                errors.append("Password must contain at least one number")
        
        if policy.get("password.require_special", {}).get("value", True):
            special_chars = policy.get("password.special_chars", {}).get("value", "!@#$%^&*()_+-=[]{}|;:,.<>?")
            if not any(c in special_chars for c in password):
                errors.append(f"Password must contain at least one special character: {special_chars}")
        
        return {
            "valid": len(errors) == 0,
            "errors": errors
        }


def get_config_service(db: Session = None) -> ConfigService:
    """Get configuration service instance."""
    if db is None:
        db = next(get_db())
    return ConfigService(db)